File: app/src/main/java/com/dm/bomber/ui/adapters/CountryCodeAdapter.java
package com.dm.bomber.ui.adapters;

import android.app.Activity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import com.dm.bomber.databinding.CountryCodeRowBinding;

public class CountryCodeAdapter extends BaseAdapter {

    private final int[] flags;
    private final String[] countryCodes;

    private final LayoutInflater inflater;

    public CountryCodeAdapter(Activity context, int[] flags, String[] countryCodes) {
        this.flags = flags;
        this.countryCodes = countryCodes;

        inflater = LayoutInflater.from(context);
    }

    @Override
    public int getCount() {
        return flags.length;
    }

    @Override
    public Object getItem(int i) {
        return null;
    }

    @Override
    public long getItemId(int i) {
        return 0;
    }

    @Override
    public View getView(int index, View view, ViewGroup parent) {
        ViewHolder holder;

        if (view == null) {
            CountryCodeRowBinding binding = CountryCodeRowBinding.inflate(inflater);

            view = binding.getRoot();
            view.setTag(holder = new ViewHolder(binding));
        } else {
            holder = (ViewHolder) view.getTag();
        }

        holder.binding.icon.setImageResource(flags[index]);
        holder.binding.code.setText(String.format("+%s", countryCodes[index]));

        return view;
    }

    private static class ViewHolder {
        CountryCodeRowBinding binding;

        public ViewHolder(CountryCodeRowBinding binding) {
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/dm/bomber/ui/adapters/CountryCodeAdapter.kt
package com.dm.bomber.ui.adapters

import android.app.Activity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import com.dm.bomber.databinding.CountryCodeRowBinding

class CountryCodeAdapter(
    context: Activity?,
    private val flags: IntArray,
    private val countryCodes: Array<String>
) : BaseAdapter() {

    private val inflater = LayoutInflater.from(context)

    override fun getCount(): Int = flags.size

    override fun getItem(i: Int): Long? = null

    override fun getItemId(i: Int): Long = 0

    override fun getView(index: Int, view: View?, parent: ViewGroup): View {
        var resultView: View? = view

        var holder: ViewHolder

        if (resultView == null) {
            val binding = CountryCodeRowBinding.inflate(inflater)
            resultView = binding.getRoot()
            resultView.setTag(ViewHolder(binding).also { holder = it })
        } else {
            holder = resultView.tag as ViewHolder
        }

        holder.binding.countryFlag.setImageResource(flags[index])
        holder.binding.countryCode.text = String.format("+%s", countryCodes[index])

        return resultView
    }

    private data class ViewHolder(val binding: CountryCodeRowBinding)
}